flow/include/flow/Traceable.h

```diff
@@ -0,0 +1,247 @@
+
+template <size_t S>
+struct TraceableString<char[S]> {
+	static_assert(S > 0, "Only string literals are supported.");
+	static const char* begin(const char* value) { return value; }
+
+	static bool atEnd(const char* value, const char* iter) {
+		return iter - value == S - 1; // Exclude trailing \0 byte
+	}
+
+	static std::string toString(const char* value) { return traceableStringToString(value, S); }
+};
+
+template <>
+struct TraceableString<char*> {
+	static const char* begin(char* value) { return value; }
+
+	static bool atEnd(char* value, const char* iter) { return *iter == '\0'; }
+
+	static std::string toString(char* value) { return std::string(value); }
+};
+
+template <class T>
+struct TraceableStringImpl : std::true_type {
+	static constexpr bool isPrintable(char c) { return 32 <= c && c <= 126; }
+
+	template <class Str>
+	static std::string toString(Str&& value) {
+		// if all characters are printable ascii, we simply return the string
+		int nonPrintables = 0;
+		int numBackslashes = 0;
+		int size = 0;
+		for (auto iter = TraceableString<T>::begin(value); !TraceableString<T>::atEnd(value, iter); ++iter) {
+			++size;
+			if (!isPrintable(char(*iter))) {
+				++nonPrintables;
+			} else if (*iter == '\\') {
+				++numBackslashes;
+			}
+		}
+		if (nonPrintables == 0 && numBackslashes == 0) {
+			return TraceableString<T>::toString(std::forward<Str>(value));
+		}
+		std::string result;
+		result.reserve(size - nonPrintables + (nonPrintables * 4) + numBackslashes);
+		int numNull = 0;
+		for (auto iter = TraceableString<T>::begin(value); !TraceableString<T>::atEnd(value, iter); ++iter) {
+			if (*iter == '\\') {
+				if (numNull > 0) {
+					result += format("[%d]", numNull);
+					numNull = 0;
+				}
+				result.push_back('\\');
+				result.push_back('\\');
+			} else if (isPrintable(*iter)) {
+				if (numNull > 0) {
+					result += format("[%d]", numNull);
+					numNull = 0;
+				}
+				result.push_back(*iter);
+			} else {
+				const uint8_t byte = *iter;
+				if (PRINTABLE_COMPRESS_NULLS && byte == 0) {
+					numNull++;
+				} else {
+					result.push_back('\\');
+					result.push_back('x');
+					result.push_back(base16Char(byte / 16));
+					result.push_back(base16Char(byte));
+				}
+			}
+		}
+		if (numNull > 0) {
+			result += format("[%d]", numNull);
+			numNull = 0;
+		}
+		return result;
+	}
+};
+
+template <>
+struct Traceable<const char*> : TraceableStringImpl<const char*> {};
+template <>
+struct Traceable<char*> : TraceableStringImpl<char*> {};
+template <size_t S>
+struct Traceable<char[S]> : TraceableStringImpl<char[S]> {};
+template <>
+struct Traceable<std::string> : TraceableStringImpl<std::string> {};
+template <>
+struct Traceable<std::string_view> : TraceableStringImpl<std::string_view> {};
+
+template <class T>
+struct Traceable<std::atomic<T>> : std::true_type {
+	static std::string toString(const std::atomic<T>& value) { return Traceable<T>::toString(value); }
```

Comment:
```suggestion
	static std::string toString(const std::atomic<T>& value) { return Traceable<T>::toString(value.load()); }
```

Reply:
The atomic provides an `operator T` that I think works in this context, but I'm happy to make it more explicit